plugins/blockvolume/hostvol_manager.go

```diff
@@ -0,0 +1,163 @@
+package blockvolume
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gluster/glusterd2/glusterd2/commands/volumes"
+	"github.com/gluster/glusterd2/glusterd2/gdctx"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/pkg/api"
+	"github.com/gluster/glusterd2/plugins/blockvolume/utils"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+const (
+	globalLockID = "host-vol-lock"
+)
+
+// HostingVolumeManager provides methods for host volume management
+type HostingVolumeManager interface {
+	GetHostingVolumesInUse() []*volume.Volinfo
+	GetOrCreateHostingVolume(name string, minSizeLimit int64) (*volume.Volinfo, error)
+}
+
+// glusterVolManager is a concrete implementation of HostingVolumeManager
+type glusterVolManager struct {
+	hostVolOpts *HostingVolumeOptions
+}
+
+// newGlusterVolManager returns a glusterVolManager instance
+func newGlusterVolManager() *glusterVolManager {
+	var (
+		g           = &glusterVolManager{}
+		hostVolOpts = &HostingVolumeOptions{}
+	)
+
+	hostVolOpts.ApplyFromConfig(viper.GetViper())
+	g.hostVolOpts = hostVolOpts
+	return g
+}
+
+// GetHostingVolumesInUse lists all volumes which used in hosting block-vols
+func (g *glusterVolManager) GetHostingVolumesInUse() []*volume.Volinfo {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	volumes, err := volume.GetVolumes(ctx)
+	if err != nil || len(volumes) == 0 {
+		return nil
+	}
+
+	return volume.ApplyFilters(volumes, volume.BlockHosted)
+}
+
+// GetOrCreateHostingVolume will returns volume details for a given volume name and having a minimum size of `minSizeLimit`.
+// If volume name is not provided then it will create a gluster volume with default size for hosting gluster block.
+func (g *glusterVolManager) GetOrCreateHostingVolume(name string, minSizeLimit int64) (*volume.Volinfo, error) {
+	var (
+		volInfo      *volume.Volinfo
+		volCreateReq = g.hostVolOpts.PrepareVolumeCreateReq()
+		ctx          = gdctx.WithReqLogger(context.Background(), log.StandardLogger())
+		clusterLocks = transaction.Locks{}
+	)
+
+	if err := clusterLocks.Lock(path.Join(globalLockID, name)); err != nil {
```

Comment:
Cluster wide lock with volume name will not help. If two concurrent requests gets different volume name then that will succeed and we end up creating two host volumes. Since CreateVolume anyways takes lock on Volume name, can we only take cluster wide lock with globalLockID only?

Reply:
I have a doubt here, if hostvolume name is present in request body, it means that volume already exists and we should not create it.
As mentioned in design doc we should only create hostvolume if hostvolume name  is not provided in req body.

Reply:
yes, I am not using the generated name, in taking lock.